### PR TITLE
Add non-broken fallback text when maxlength/maxwords is 0

### PR DESCRIPTION
**Pull request by romaricpascal**
Avoids displaying a "You can enter **undefined** characters" to users if neither `maxlength` nor `maxwords` is set.

This'll avoid services from getting an unsavory message in front of their users if they chose to use JavaScript to pass the `maxlength` or `maxcount` option. (Closes https://github.com/alphagov/govuk-frontend/issues/2910).

Unfortunately, before JavaScript kicks in, we can't tell if it'll be characters or words (nor the limit obviously), so the message is pretty vague (but still better than having a message that looks like there was a bug). @claireashworth, if you have a better wording, I'm open to suggestions.

Given we're expecting JavaScript to kick in and set a dynamic text based on the content of the component, we do need to have an element on the page taking some vertical space to avoid layout shifts, so some content is required. And as we're passing the option as `text`, `&nbsp;` or other kind of HTML entities will be escaped.

### CODE REVIEW

**Comment by claireashworth** [comment]
I agree. If there's no maxlength or maxword defined is there a default limit or does it become unlimited? If there's a default then we can refer to that. Is having nothing defined done for a particular use case? If so, we could possibly reference that. Otherwise, I agree it would be better if there was no message.

**Comment by claireashworth** [comment]
In that case, no message sounds like the better option.

**Comment by 36degrees** [comment]
I think an unfortunate side effect of the config work that we didn't foresee is that the Character Count JS can now receive config which isn't available to the template, which I think needs to be factored in here.

I'd agree with omitting the count message entirely if neither `maxwords` nor `maxlength` is set **in the macro**, but if the component is then initialised in JS with `new CharacterCount($element, { maxlength: 200 })` it feels like that should still work – there just wouldn't be an existing 'fallback message' in that context, and the JS might need to allow for that?

**Comment by romaricpascal** [comment]
That hint element does two nice little things for the component (besides serving as an anchor for the JavaScript, but we could easily default to the textarea being that anchor if there's no element):

1. it reserves space on the page, which means that when JavaScript will kick in, there won't be a shift of content. I'm not sure how critical that is to maintain that, it would provide a better experience for the users if buttons don't shift around as JavaScript loads.
2. it serves as `aria-describedby` for the textarea, letting users know that the length of the input is limited when they focus the field. We don't have a number until JavaScript kicks in and when JavaScript does kick in, we have no string to format to inject it. My understanding is that we set the `aria-describedby` to that element to avoid double announcements (based on what I gathered from the issues linked on https://github.com/alphagov/govuk-frontend/pull/2577. We could use try passing down the fallback hint down to JavaScript for use as a translation:
 - if no `maxwords` or `maxlength` is set, pass it as `data-i18n.fallback-hint`, JS picks it up and injects it into the element. We don't pluralise in Nunjucks so we could not worry about it there as well (and it'll likely be a large number anyway)
 - if `maxwords` or `maxlength` are set, interpolate like we do now and don't pass it to JS as it'll be on screen.
 
Feels like quite a bit of impact though, I may be missing a simpler route 😬

**Comment by querkmachine** [comment]
I might be misunderstanding here, sorry.

The primary purpose of the fallback hint is so that guidance on character limits is still available in situations where JavaScript has been disabled or failed to load/initialise correctly. In that regard, anything that requires passing it to or processing it in JavaScript is going to fail that requirement.

**Comment by romaricpascal** [comment]
No worries, I might be not explaining it super well as well 😊 .

The element does indeed act as the fallback visible text before JavaScript loads (I should have listed that as the first thing it does). And for that role, I agree with you, it doesn't make sense to be relying on JavaScript to inject any content.

However, the element also serves to provide guidance on the character limit for assistive technology, through being linked to the `textarea` via `aria-describedby`. If it's empty and we don't fill it in when the component is configured in JavaScript, users of services chosing that road for configuring their `CharacterCount` won't have an announcement that there's a limit to the number of characters when they focus the field, only announcements when they type and the number of remaining/over limit characters get updated. 

That's for this scenario that I propose to pass the `fallbackHintText` (or `fallbackHintDefault` if none is set) as a `data-i18n....` attribute so that the JavaScript can set that accessible description, replacing `%{count}` with whatever was configured at initialisation.

The lack of `maxlength`/`maxwords` in the macro options would trigger the template to switch from injecting `fallbackHintText`/`fallbackHintDefault` inside the HTML and instead pass it as an i18n message to be set at component initialisation.

**Comment by querkmachine** [comment]
Ahh, I think I understand now.

It's not the way we would want service teams to be doing this (we'd expect one of `maxlength` or `maxwords` to be set in the Nunjucks in virtually all circumstances, for progressive enhancement as much as accessibility), but on the offchance that a JS configuration is present and a Nunjucks one is not, then populating the fallback dynamically makes sense. 👍

File: src/govuk/components/character-count/character-count.test.js [hunk 1]
```diff
@@ -433,6 +433,28 @@ describe('Character count', () => {
           const visibility = await page.$eval('.govuk-character-count__status', el => window.getComputedStyle(el).visibility)
           expect(visibility).toEqual('visible')
         })
+
+        it.only('injects the description of the textarea', async () => {
```
[the remark this answers is not in the document]
Comment (romaricpascal) [reply src/govuk/components/character-count/character-count.test.js]: Whoops, had forgotten to clean that up 😬 Cheers for spotting it, should be all tidy now (also removed the `console.log`ging that was set up in the test).

File: src/govuk/components/character-count/character-count.mjs [hunk 2]
```diff
@@ -120,6 +120,13 @@ CharacterCount.prototype.init = function () {
   var $textarea = this.$textarea
   var $fallbackLimitMessage = document.getElementById($textarea.id + '-info')
 
+  // Inject a decription for the textarea if none is present already
+  // for when the component was rendered with no maxlength, maxwords
+  // nor custom fallbackHintText
+  if ($fallbackLimitMessage.textContent.match(/^\s*$/)) {
+    $fallbackLimitMessage.textContent = this.i18n.t('fallbackHint', { count: this.maxLength })
```
Comment (36degrees) [inline src/govuk/components/character-count/character-count.mjs:127]: If folks aren't using the Nunjucks macro, and don't have the corresponding changes to pass us the fallback hint, this could result in an error being thrown.

Given we're already in a 'best efforts' context trying to do something sensible, I think we should avoid that by guarding this – either by checking for `fallbackHint` being set or just wrapping this in a try catch?

File: src/govuk/components/character-count/character-count.test.js [hunk 3]
```diff
@@ -433,6 +433,21 @@ describe('Character count', () => {
           const visibility = await page.$eval('.govuk-character-count__status', el => window.getComputedStyle(el).visibility)
           expect(visibility).toEqual('visible')
         })
+
+        it('injects the description of the textarea', async () => {
```
Comment (36degrees) [inline src/govuk/components/character-count/character-count.test.js:437]: I think this test description needs more context – outside of the context of this PR if this test started failing I'm not sure I'd understand what it was trying to test.

File: src/govuk/components/character-count/character-count.mjs [hunk 4]
```diff
@@ -367,14 +374,11 @@ export default CharacterCount
 
 /**
  * @typedef {object} CharacterCountTranslations
- * @property {PluralisedTranslation} [charactersUnderLimit] - Characters under limit
+ * @property {import('../../i18n.mjs').PluralisedTranslation} [charactersUnderLimit] - Characters under limit
```
Comment (36degrees) [inline src/govuk/components/character-count/character-count.mjs:377]: Are these changes related?

File: src/govuk/components/character-count/template.test.js [hunk 5]
```diff
@@ -260,4 +260,23 @@ describe('Character count', () => {
       })
     })
   })
+
+  describe('no maximum', () => {
```
Comment (36degrees) [inline src/govuk/components/character-count/template.test.js:264]: Again this could do with being clearer, maybe something like:

```suggestion
  describe('when neither maxlength nor maxwords are set', () => {
```

File: src/govuk/components/character-count/template.test.js [hunk 6]
```diff
@@ -260,4 +260,23 @@ describe('Character count', () => {
...
+
+      // Fallback hint is passed as data attribute
+      const $component = $('[data-module]')
+      expect($component.attr('data-i18n.fallback-hint.other')).toEqual('No more than %{count} characters')
+
+      // A non-breaking space is set as the count message to reserve space
+      // and prevent a layout shift when the JavaScript component
+      // initialises and fills the element with the relevant text
+      const $countMessage = $('.govuk-character-count__message')
+      expect($countMessage.html()).toContain('&#xA0;')
```
Comment (36degrees) [inline src/govuk/components/character-count/template.test.js:279]: Worth rebasing against main – suspect this might fail once rebased on top of 7bbc33234ad431e88e526f1b698355022e68e74d.

File: src/govuk/components/character-count/template.njk [hunk 7]
```diff
@@ -3,10 +3,14 @@
 {% from "../textarea/macro.njk" import govukTextarea %}
 {% from "../hint/macro.njk" import govukHint %}
 
+{%- set hasNoLimit = (not params.maxwords and not params.maxlength) %}
+{%- set fallbackHintText = params.fallbackHintText or 'You can enter up to %{count} ' + ('words' if params.maxwords else 'characters') %}
```
Comment (36degrees) [inline src/govuk/components/character-count/template.njk:7]: If neither maxwords nor maxlength are passed this will always end up talking about characters, but `maxwords` could then be used when the component is initialised in JS.

You can see this if you go to https://govuk-frontend-pr-2915.herokuapp.com/components/character-count/no-maximum/preview and then run `new GOVUKFrontend.CharacterCount($('[data-module]'), { maxwords: 10 }).init()` in the console. The visible message will correctly display the number of words remaining, but if you inspect the fallback hint will still be talking about characters.

Unfortunately I'm not sure that's really fixable with the current approach without making things quite a lot messier… 🤔

File: src/govuk/components/character-count/template.njk [hunk 8]
```diff
@@ -35,9 +39,14 @@
...
-  {%- set fallbackHintDefault = 'You can enter up to %{count} ' + ('words' if params.maxwords else 'characters') %}
+  {#
+    Use the fact that `html` has precedence over `text` to override content
+    when no limit is set and reserve vertical space to avoid layout shifting
+    by injecting a non-breakable space
+  #}
   {{ govukHint({
-    text: (params.fallbackHintText or fallbackHintDefault) | replace('%{count}', fallbackHintLength),
+    text: ((fallbackHintText) | replace('%{count}', fallbackHintLength) if not hasNoLimit),
+    html: ('&nbsp;' if hasNoLimit),
```
Comment (36degrees) [inline src/govuk/components/character-count/template.njk:49]: This does feel a little messy. Given this is (hopefully) something of an edge case, I'm wondering if we just live with the layout shift as the hint gets injected.

Alternatively, what if we added a class to make the hint `visibility: hidden` if we don't know the count, then remove it once the JS fires?

File: src/govuk/components/character-count/character-count.mjs (hunk 2 above)
Comment (romaricpascal) [reply src/govuk/components/character-count/character-count.mjs:127]: I don't think anything will be thrown there:
- `$fallbackLimitMessage` was already used beforehand, it's the reference for where to insert the visible and screen-reader containers
- the I18n class will not throw if there's no entry for `fallbackHint` in its content, it [returns the lookup key if no translation was provided](https://github.com/alphagov/govuk-frontend/blob/main/src/govuk/i18n.mjs#L43-L61). 

This would set a default of `fallbackHint` as a description for assistive technologies. How about we provide our own default of `{other: ''}` to ensure we have a smooth experience on that side? That'll set an empty string if no alternative is provided by data attributes (or Nunjucks) and warn people in the console if they needed to use any of the other keys.

Thinking about it, we should probably warn in I18n if no key is found for the translation as well.

Comment (36degrees) [reply src/govuk/components/character-count/character-count.mjs:127]: Because we're treating `fallbackHint` as a pluralised string, we'll call `getPluralSuffix` before we get to that code, and  `getPluralSuffix` will throw if `fallbackHint.other` is not defined.

https://github.com/alphagov/govuk-frontend/blob/5e0746848368760a1a263057b4957369e922a82d/src/govuk/i18n.mjs#L162-L179

> How about we provide our own default of {other: ''} to ensure we have a smooth experience on that side?

That could work 👍🏻

File: src/govuk/components/character-count/character-count.mjs (hunk 4 above)
Comment (romaricpascal) [reply src/govuk/components/character-count/character-count.mjs:377]: Good spot, that was a poor conflict resolution from rebasing `main`.

File: src/govuk/components/character-count/template.test.js (hunk 6 above)
Comment (romaricpascal) [reply src/govuk/components/character-count/template.test.js:279]: `main` is already rebased. Because this is a non breaking space, Cheerio still encodes it (though into its hexa decimal form rather than the initiaal `&nbsp;`).

Comment (36degrees) [reply src/govuk/components/character-count/template.test.js:279]: According to GitHub this branch is currently 26 commits behind main, including 7bbc33234ad431e88e526f1b698355022e68e74d.

But happy that it might not be an issue anyway 👍🏻

File: src/govuk/components/character-count/template.njk (hunk 7 above)
Comment (romaricpascal) [reply src/govuk/components/character-count/template.njk:7]: Yeah, we can only do so much without knowing if it's characters or words that will be counted. In the first take I had set a very generic "You can enter a limited amount of content". Discussing with Beeps and Claire, it doesn't bring anything to the table so left it out.

I may have a tidy way to sort that out: leave the attribute out if no fallback text is set and no maxwords/maxlength are (but keep injecting the default as we do in the HTML from Nunjucks if we have a `maxlength` or `maxwords`). With a default of `''` in the JS, that'd mean no description for the user when they first focus the field, then an announcement as they reach the threshold and the count message starts displaying. This sounds like the tidiest scenario for users.

File: src/govuk/components/character-count/template.test.js (hunk 6 above)
Comment (romaricpascal) [reply src/govuk/components/character-count/template.test.js:279]: Yup, I'd forgotten to push the rebasing, so got confused as to why things were still marked as OK on Github 🤦🏻

**Comment by romaricpascal** [comment]
@36degrees All your comments should have been adressed in that last push 😄 Main changes, as discussed, were:
- moving to not passing a text to JavaScript unless people provide an explicit fallback hint, as we cannot guess whether it'll be characters or words
- providing a default translation for the fallbackHint's other (hadn't thought that this would throw indeed, good spot)
- vertical space is now reserved in CSS rather than in the markup. Still with extra content, but a zero-width space rather than non-breaking one (that allow us to have it at all times, which is simpler).

**Comment by romaricpascal** [comment]
How would we feel about renaming that option "textareaDescriptionText" (`data-i18n.textarea-description`, `i18n.textareaDescription`) so we don't miss that it provides the textarea description (and just mention that this description will be visible to users until the JavaScript component is initialised)?

**Comment by 36degrees** [comment]
> How would we feel about renaming that option "textareaDescriptionText" (`data-i18n.textarea-description`, `i18n.textareaDescription`) so we don't miss that it provides the textarea description (and just mention that this description will be visible to users until the JavaScript component is initialised)?

Sounds good to me.

Changes look good too – but if we're planning to rename then I'll hold off approving until that's done.

**Review by 36degrees — COMMENTED** [review]
Looks like there are still quite a lot of places where we talk about a 'fallback message' or 'fallback hint' - suspect some of them are fine but think it might be worth updating at least `$fallbackLimitMessage` and the related comments in the component JS?

```
$ ag fallback src/govuk/components/character-count 
src/govuk/components/character-count/character-count.yaml
43:  description: Text describing the maximum number of characters you can enter, which is announced to screen readers. The text is displayed as a fallback if the character count JavaScript does not run. Depending on how you configure this component and what parameters you add, instances of `%{count}` are replaced by the value of `maxwords`. If not configured, it uses `maxlength`. By default, fallback text is provided in English.
89:  - name: with custom fallback text
90:    description: with no-js fallback text translated into Welsh
92:      name: custom-fallback
93:      id: custom-fallback
300:  - name: when neither maxlength/maxwords nor fallback hint are set

src/govuk/components/character-count/template.njk
14:  This means we can't compute a default fallback message (also used as the textarea accessible

src/govuk/components/character-count/character-count.test.js
34:    it('shows the fallback message', async () => {
58:      it('hides the fallback hint', async () => {
440:                'when neither maxlength/maxwords nor fallback hint are set'
609:        it('interpolates the fallback hint in data attributes with the maximum set in JavaScript', async () => {
610:          // This tests that any fallback hint provided through data-attributes

src/govuk/components/character-count/character-count.mjs
94:    // Read the fallback if necessary rather than have it set in the defaults
124:  var $fallbackLimitMessage = document.getElementById($textarea.id + '-info')
129:  if ($fallbackLimitMessage.textContent.match(/^\s*$/)) {
130:    $fallbackLimitMessage.textContent = this.i18n.t('textareaDescription', { count: this.maxLength })
133:  // Move the fallback count message to be immediately after the textarea
135:  $textarea.insertAdjacentElement('afterend', $fallbackLimitMessage)
143:  $fallbackLimitMessage.insertAdjacentElement('afterend', $screenReaderCountMessage)
146:  // fallback element for backwards compatibility as these may have been
149:  $visibleCountMessage.className = $fallbackLimitMessage.className
153:  $fallbackLimitMessage.insertAdjacentElement('afterend', $visibleCountMessage)
155:  // Hide the fallback limit message
156:  $fallbackLimitMessage.classList.add('govuk-visually-hidden')
386: * @property {PluralisedTranslation} [textareaDescription] - Fallback hint

src/govuk/components/character-count/template.test.js
228:  describe('with custom fallback text', () => {
229:    it('allows customisation of the fallback message', () => {
230:      const $ = render('character-count', examples['with custom fallback text'])
261:    describe('with fallback hint set', () => {
263:      // it needs to pass down any fallback hint to the JavaScript
265:      it('renders the fallback hint as a data attribute', () => {
268:        // Fallback hint is passed as data attribute
279:    describe('without fallback hint', () => {
280:      it('does not render a fallback hint data attribute', () => {
283:          examples['when neither maxlength/maxwords nor fallback hint are set']
```

**Comment by romaricpascal** [comment]
Yup, that last commit just updated the full variable name to its new one. I'm starting a round of the rest of the code to check for further references to "fallback" 😄

**Comment by romaricpascal** [comment]
@36degrees  I _think_ I've shifted all references of "fallback" to "textarea description" when talking about the Character Count (appart from when saying that this description is a fallback when JS doesn't run) 🎉 I'll update the documentation draft as well 😄

**Review by 36degrees — APPROVED** [review]
